### ISSUE TITLE
Cleaning up when directory connection fails

### ISSUE THREAD

**Issue by dannylindquist** [Abc-Arbitrage/Zebus#82]
Hey team,

I am trying to create some resiliency when being disconnected from the Directory Service. I know that using multiple instances with cassandra already helps with this matter, but I'm trying to attempt re-connection to a single instance of the directory in the case it goes down or a server needs a reboot. My scenario is to attempt re-connection periodically by trying to create a new bus that connects to the directory, but then tears it down if it can't register. Even with a `using` statement, it appears the sockets behind the scenes are still up. The dispose function fails if a peer cannot be unregistered. 
Is there a way I could get around this? Or maybe the dispose could succeed regardless or whether or not a unregister is sucessful?

Thanks in advance

**Comment by ocoanet** [comment]
Hello.

I cannot reproduce the connection leaks using the following simple code:
```cs
private static IBus StartBus(BusFactory busFactory, int maxAttempts, TimeSpan attemptDelay)
{
    var attemptCount = 0;
    while (true)
    {
        Console.WriteLine($"Trying to start bus, attempt #{attemptCount}");

        if (TryStartBus(busFactory, out var bus))
            return bus;

        if (++attemptCount >= maxAttempts)
            throw new TimeoutException("Unable to start bus");

        Console.WriteLine($"Unable to start bus, retrying in {attemptDelay}");

        Thread.Sleep(attemptDelay);
    }
}

private static bool TryStartBus(BusFactory busFactory, out IBus bus)
{
    bus = busFactory.CreateBus();
    try
    {
        bus.Start();

        return true;
    }
    catch (Exception)
    {
        bus.Dispose();
        bus = default;

        return false;
    }
}
```
Can you provide a short, working example?

**Comment by ocoanet** [comment]
I like the idea to provide auto-retry options for Directory commands. This could be easily implemented in Zebus using additional parameters like "RegistrationRetryAttempts" and "RegistrationRetryDelay". However it would also require to add a `Task IBus.StartAsync()` method to make explicit the async nature of the registration process.

**Comment by dannylindquist** [comment]
I can confirm that your code works perfectly on my end as well. I apologize, it turns out I was testing my code originally in the Zebus.Samples project which was using `v1.3.3`. I updated and everything seems to be running as expected. Thanks for helping me realize that.

I do think the `IBus.StartAsync` could be a very useful alternative though. We would certainly use it if it was there. There is no rush since the other code works too, but having it build it would be a great feature.

Thanks again for your helpful response.

**Comment by dannylindquist** [comment]
Hey team,
Sorry to post on a closed issue. The issue happens when you connect to a Directory, then the directory quits, and then you attempt to dispose the bus instance. It will continue sending a sync message to the directory indefinitely until the processed is closed.

Is there a way around this.

To replicate, run the stand alone directory in one process, then run the following code. You will see the underlying ports will continue to stay open. It will even reconnect after disposing if you restart the service. 

```csharp
static void Main(string[] args) {
    var busfactory = new BusFactory().WithScan().WithConfiguration("tcp://localhost:129", "Demo");
    var bus = StartBus(busfactory, 10, TimeSpan.FromSeconds(5));
    Task.Delay(30000).Wait(); // give bus and directory a chance to chat
    System.Diagnostics.Process.GetProcessesByName("Abc.Zebus.Directory")[0].Kill(); //Stop the directory
    Task.Delay(3000).Wait(); 
    try {
        bus.Dispose();
    } catch {
        //fail to unregister
    }
    Console.ReadLine(); //never gets cleaned up
}

private static IBus StartBus(BusFactory busFactory, int maxAttempts, TimeSpan attemptDelay) {
    var attemptCount = 0;
    while (true) {
        Console.WriteLine($"Trying to start bus, attempt #{attemptCount}");

        if (TryStartBus(busFactory, out var bus))
            return bus;

        if (++attemptCount >= maxAttempts)
            throw new TimeoutException("Unable to start bus");

        Console.WriteLine($"Unable to start bus, retrying in {attemptDelay}");

        Thread.Sleep(attemptDelay);
    }
}

private static bool TryStartBus(BusFactory busFactory, out IBus bus) {
    bus = busFactory.CreateBus();
    try {
        bus.Start();

        return true;
    } catch (Exception) {
        bus.Dispose();
        bus = null;

        return false;
    }
}
```

**Comment by ltrzesniewski** [comment]
Hi. Thanks for the report.

If the bus fails to unregister from the directory, it throws the *"Unable to unregister peer on directory"* exception before shutting down its transport, which causes the reconnect attempts you've observed. The bus is then stuck in the `Stopping` state forever.

We should shut down the bus properly when this happens, but I'm not sure if we should rethrow the exception, as the `Stop` method caused the bus to stop after all.

**Comment by ocoanet** [comment]
Yes, I agree the `Bus.Stop` issue must be fixed. @ltrzesniewski Do you have the time to handle it?

@dannylindquist please note that our current bus usage is to start the bus when a process (service, client) starts, and to stop the bus when the process stops. This usage has a few consequences:
- Starting the bus is critical because our services cannot start without a running bus. Thus, we have 3 Directories per environment (with a Cassandra storage) to make sure there is always a Directory running to handle registration and unregistration commands.
- The fact that we have multiple Directories per environment explains why there is not auto-retry on registration commands.
- The bus should be restartable, i.e., it should be possible to start a bus instance again after stop. However, I don't think this ability is often used in our production components, so you might discover issues if you use it.

On a side note, I do not think using an in-memory Directory is acceptable for a production scenario. You should at least use a file based Directory, using a storage engine like RocksDB. We have wanted to add a RocksDB based Directory to provide a production-ready alternative to Cassandra, but this has not been planned yet.

**Comment by ltrzesniewski** [comment]
@dannylindquist This should be fixed in v3.2.2.
